lib/container/ceph_container.ex

```diff
@@ -128,6 +136,14 @@ defmodule Testcontainers.CephContainer do
     %{config | wait_timeout: wait_timeout}
   end
 
+  @doc """
+  Set the regular expression to check the image validity.
+  """
+  def with_check_image(%__MODULE__{} = config, check_image)
```

Comment:
I might be wrong, but do we need all these methods in each of the container implementations? Can't we just call the method in Container?

Reply:
we have the methods new and with_environment_foo etc that are called from the container implementations. We can do the same with with_check_image. And it can take both string and regex. But should crash I think, what do you think

Reply:
That is the point that is eating my brain: how to avoid spreading this concern to all other modules. But if we set the `check_image` only on the `Testcontainers.ContainerBuilder.build/1` implementation, as we do for the `environment` variables, the end user won't be able to customize the validation without creating its own container.

For example, if I wanted to use the [`valkey`][0] [image][1] instead of the `redis`, that wouldn't be possible without changing the `check_image` validation:

```elixir
alias Testcontainers.RedisContainer
alias Testcontainers.ContainerBuilder
config = RedisContainer.new() |> RedisContainer.with_image("valkey/valkey:unstable")
container = ContainerBuilder.build(config)  ## raises ArgumentError
```

What do you think about moving the validation call to the private method [Testcontainers.start_and_wait/2][2]? This would give the opportunity to the end user to convert the container implementation into the generic configuration and change the validation.

[0]: https://valkey.io/
[1]: https://hub.docker.com/r/valkey/valkey
[2]: https://github.com/joaodubas/testcontainers-elixir/blob/c447aba377b98ae3ca7267a5f3921479890f129e/lib/testcontainers.ex#L225-L241

Reply:
checking out your fork  and branch to take a better look

Reply:
Ok. Looked at it. 

First, we can create a defguard in Testcontainers.Container like this

```elixir
  defguard is_valid_image(image) when is_binary(image)
    or is_struct(image, Regex)
```
Then  we import it and use it in the specific container impls:

```elixir
  import Testcontainers.Container, only: [is_valid_image: 1]
  
  # .......
  
 
  def with_check_image(%__MODULE__{} = config, check_image) when is_valid_image(check_image) do
    %__MODULE__{config | check_image: check_image}
  end
```
 
A string could also work as a regex. If you supply a string to Regex.compile it will definitely work, right?. So we should also remove the default wildcarding and just change the first method to:
 
```elixir
  def with_check_image(%__MODULE__{} = config, check_image) when is_binary(check_image) do
    regex = Regex.compile!(check_image) do
    with_check_image(config, regex)
  end
 ```
 
This will leave it all up to the author of the container to decide, both for the preset but also for the custom containers.

Next, about moving validation. I don't think it's needed. If you make a RedisContainer, you can just call with_check_image and then build:

```elixir
alias Testcontainers.RedisContainer
alias Testcontainers.ContainerBuilder
config = RedisContainer.new() |> RedisContainer.with_image("valkey/valkey:unstable") |> RedisContainer.with_check_image("valkey/valkey") # provided forward slash works in a regex
container = ContainerBuilder.build(config)  ## should not raise ArgumentError
```

Then to the last point, you shouldn't need to build anything. When you have a container like RedisContainer or a generic custom container the way you like it you just pass it into `Testcontainers.start_container´ and that method will build it for you

Reply:
check if my ramblings make sense :) had some minutes after lunch now ;)

Reply:
They make a lot of sense, reduce code duplication is always essential. I made some commits addressing these points.

As for the code duplication related to the `with_check_image` method, I was thinking about using the same pattern employed in the library [`Req][0], more specifically, the [`Req.Steps`][1]. Something like registering functions that change the container configuration and applying them at a later stage, probably before executing the container. But this means a major refactor on the code base, which can be done in a separate PR if it makes sense.

[0]: https://hex.pm/packages/req
[1]: https://github.com/wojtekmach/req/blob/15cf6dfd24309993254a8698099e92a6c33690f9/lib/req/steps.ex

Reply:
fix the conclict and we can run the tests and see if everything works :) i made a patch release after you opened the pr and there is some new network mode option going on now.

Reply:
Fixed the conflict on `lib/container.ex` and also made two more updates:

1. Add an example on `README.md`.
2. Expose the `check_image` option on `Testcontainers.Ecto`.